pkg/controller/localvolumediscovery/localvolumediscovery_controller.go

```diff
@@ -0,0 +1,219 @@
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("localvolumediscovery-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource LocalVolumeDiscovery
+	return c.Watch(&source.Kind{Type: &localv1alpha1.LocalVolumeDiscovery{}}, &handler.EnqueueRequestForObject{})
+}
+
+// blank assignment to verify that ReconcileLocalVolumeDiscovery implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileLocalVolumeDiscovery{}
+
+// ReconcileLocalVolumeDiscovery reconciles a LocalVolumeDiscovery object
+type ReconcileLocalVolumeDiscovery struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a LocalVolumeDiscovery object and makes changes based on the state read
+// and what is in the LocalVolumeDiscovery.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileLocalVolumeDiscovery) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling LocalVolumeDiscovery")
+
+	// Fetch the LocalVolumeDiscovery instance
+	instance := &localv1alpha1.LocalVolumeDiscovery{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	diskMakerDSMutateFn := getDiskMakerDiscoveryDSMutateFn(request, instance.Spec.Tolerations, getOwnerRefs(instance), instance.Spec.NodeSelector)
+	ds, opResult, err := nodedaemon.CreateOrUpdateDaemonset(r.client, diskMakerDSMutateFn)
+	if err != nil {
+		instance.Status.Phase = localv1alpha1.DiscoveryFailed
+		err = r.updateStatus(instance)
+		if err != nil {
+			reqLogger.Error(err, "failed to update status")
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, err
+	} else if opResult == controllerutil.OperationResultUpdated || opResult == controllerutil.OperationResultCreated {
+		reqLogger.Info("daemonset changed", "daemonset.Name", ds.GetName(), "op.Result", opResult)
+	}
+
+	instance.Status.Phase = localv1alpha1.Discovering
+	err = r.updateStatus(instance)
```

Comment:
We should set Conditions too.

Reply:
Added conditions and updated the unit tests to test conditions.

Reply:
and `ObservedGeneration`

Reply:
added `ObservedGeneration` in the status